### PR TITLE
Dev/version 2.0

### PR DESCRIPTION
**Pull request by KinoxKlark**
Thanks to the suggestion of @connorjak https://github.com/pyimgui/pyimgui/issues/237#issuecomment-956409058, here is a draft PR for the next major release.

This version will mainly support the upgrade to *DearImGui* v1.82 (see #192).

As I recall it there are still two main flaws with this dev branch
1. In order to implement the `internal` submodule we included the cpp lib twice which is suboptimal. One shared library would be more efficient and could open the way to external modules implementing add-ons like *implot* and *imnode*. This should be carefully studied in order to avoid future breaking changes. (See #229, https://github.com/KinoxKlark/pyimgui/pull/2)
2. The mapping of the `Table` functionalities wasn't robust as I wanted it. It may be enhanced with python specific syntax and should be reworked a bit.

There are probably many other things that I forgot. Feel free to add to the list and discuss all of this.

(Moreover, there is some PR for the docking branch implementation which seems to be a recurring asked feature. What should we do about that? #242, https://github.com/KinoxKlark/pyimgui/pull/4)

### CODE REVIEW

**Comment by mCodingLLC** [comment]
Personally I'm not worried about the inefficiency in (1.) so long as it doesn't break anything fundamental later if changed. I don't have much opinion about (2.) but I'm super excited to have docking merged into at least the dev branch.

**Comment by Inventor-Mentor** [comment]
Regarding your point 1 above, personally in my opinion the support of **implot** functionality in a 2.0 development line is a need-to-have feature for many of my own use cases. Furthermore, supporting imnode features would also be nice-to-have in the long run, even from your more focused point of view (following more closely the successful Unix philisophy “do one thing and do it well [and play nicely with others if needed]). Despite Donald Knuth’s famous quote that "premature optimization is the root of all evil" (with respect to over-engineering of details omitting the system enginnering focus that “a GUI system like pyimgui should firstly fulfill correctly and adequately its main specification”, from an engineering point of view, I would personally agree that giving careful thought to major design decisions (such as introducting shared libraries / DLLs) that might affect bottlenecks for an immediate mode rendering loop) is a very reasonable thing to do. A web search would let me to assume that such a design decision should not be too bad for many cases (primarily, only on the very first call to a shared  [dynamic-link] lilbrary), nevertheless there might be a small remaining performance penality regarding to some experiments below,, please cf.:

[https://stackoverflow.com/questions/140061/when-to-use-dynamic-vs-static-libraries](https://stackoverflow.com/questions/140061/when-to-use-dynamic-vs-static-libraries)

[https://stackoverflow.com/questions/4667882/is-a-statically-linked-executable-faster-than-a-dynamically-linked-executable](https://stackoverflow.com/questions/4667882/is-a-statically-linked-executable-faster-than-a-dynamically-linked-executable)

[https://gamedev.net/forums/topic/360353-dll-performance/3367636/
](https://gamedev.net/forums/topic/360353-dll-performance/3367636/)

[https://gcc.gnu.org/legacy-ml/gcc/2004-06/msg01956.html](https://gcc.gnu.org/legacy-ml/gcc/2004-06/msg01956.html)

In my opinion, correctness regarding the overall specification and philiosophy of pyimgui, should be fulfilled first and foremost also in the 2.0 development line. Nevertheless, a performance bug could very well be a mistake in an immediate mode library at least from the point of view of more demanding dynamic use cases especially on weaker embadded hardware like the Raspberry Pi. Therefore, at some point some performance experiments or even better automatic performance regression tests would be a reasonable thing to have from an engineering point of view employing cross-language-barrier performance profilers spanning Python / C / C++ / Fortan code and other linked in binary machine code [numpy / scipy encapsulating machine code from FORTRAN libraries such as LAPACK] such as **Scalene** [Open and Free] or Intel’s **vTune** [closed and now free to use, however, on Intel architecture only, not available on ARM / Raspberry Pi ] to tell if there are actually performance penalties employing DLLs and what current performance bottlenecks actiually are, please cf..:

[https://github.com/plasma-umass/scalene](https://github.com/plasma-umass/scalene)

Further infos (lecture video and slides) on Scalene are available here:

[https://learning.acm.org/techtalks/performance](https://learning.acm.org/techtalks/performance)

[https://www.intel.com/content/www/us/en/developer/tools/oneapi/vtune-profiler.html
](https://www.intel.com/content/www/us/en/developer/tools/oneapi/vtune-profiler.html)

Just my 2 cents. Hope they are of any help! Have a nice weekend and best regards and wishes, from Peter

**Comment by KinoxKlark** [comment]
> Regarding your point 1 above, personally in my opinion the support of **implot** functionality in a 2.0 development line is a need-to-have feature for many of my own use cases. Furthermore, supporting imnode features would also be nice-to-have in the long run, even from your more focused point of view (following more closely the successful Unix philisophy “do one thing and do it well [and play nicely with others if needed]). Despite Donald Knuth’s famous quote that "premature optimization is the root of all evil" (with respect to over-engineering of details omitting the system enginnering focus that “a GUI system like pyimgui should firstly fulfill correctly and adequately its main specification”, from an engineering point of view, I would personally agree that giving careful thought to major design decisions (such as introducting shared libraries / DLLs) that might affect bottlenecks for an immediate mode rendering loop) is a very reasonable thing to do. [...]

Thank you very much for your help and insights. Those are interesting readings that you share here. Following what you discussed here, I created a `feature/dll` branch (#247) from which I hope to set up a version of `pyimgui` that depends only on a dynamically linked version of *DearImGui*. This branch should serve as a strong basis for this refactor and be the starting point of performance testing. It may also serve as a basis for a prototype version of `pyimnode`, permitting us to understand what is needed for extern developers to create addons to `pyimgui`.

File: config-cpp/py_imconfig.h [hunk 1]
```diff
@@ -40,3 +40,6 @@ void __py_assert(const char* msg);
 
 // tune vertex index size to enable longer draw lists (see #138 issue)
 #define ImDrawIdx unsigned int
+
+//---- Use 32-bit for ImWchar (default is 16-bit) to support unicode planes 1-16. (e.g. point beyond 0xFFFF like emoticons, dingbats, symbols, shapes, ancient languages, etc...)
+#define IMGUI_USE_WCHAR32
```
Comment (learn-more) [inline config-cpp/py_imconfig.h:45]: UTF16 also supports those planes, but it just uses 2 or 3 characters per code point?.

Comment (KinoxKlark) [reply config-cpp/py_imconfig.h:45]: I think this comes straight from DearImGui comment documentation of `imconfig.h` (https://github.com/ocornut/imgui/blob/43177324c081868116a99f634e2d5bf38b54056b/imconfig.h#L58), but I must admit I am not at all familiar with character encoding specifications and I just don't know if this is needed or not. Do you suggest that we don't need to apply this change? (this was from pr #241)

Comment (learn-more) [reply config-cpp/py_imconfig.h:45]: after re-reading this, the support indicated by the comment is in DearImGui, not in unicode.
DearImGui puts all characters in a dictionary or set or whatever, and that is a whole lot simpler with wchar32's ridiculous big characters.

Comment (KinoxKlark) [reply config-cpp/py_imconfig.h:45]: Ok so, if I understand correctly, we need to keep wchar32 in accordance with DearImGui? (i.e. keep this change)

Comment (learn-more) [reply config-cpp/py_imconfig.h:45]: indeed.

**Comment by KinoxKlark** [comment]
As suggested in #299, it may be a good time to release the version 2.0 as it is.

I didn't manage to find a proper solution for the DLL feature and as it is DearImGui is statically linked twice (in `imgui.core` and in `imgui.internal`). This is not ideal but either absolutely problematic, and it is indeed better to release something as it is than waiting one eternity that I could find some time to work on this feature.

At least the documentation now builds correctly.

I plan to merge bug fix #296 before merging `dev/version-2.0` into `master`. I will also consider merging PR relative to bug fix in coming days.

I thus won't merge following features for now (but will be considered for future minor releases):
- #247
- #272

Please, if you think I missed something important that should be done before releasing 2.0 let me know. 

Once 2.0 will be released, I propose to update to a more recent version of DearImGui and focus on the `docking` branch.

**Comment by bernd-schotanus** [comment]
Please merge this PR. version 2.0 is required to get pyimgui up and running with pygame, it should be the default.

**Pull request merged**